Source/WebCore/dom/Element.cpp

```diff
@@ -5107,8 +5107,43 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
 #else
     bool optionsEnabled = document().settings().fullScreenKeyboardLock();
 #endif
-    if (optionsEnabled && document().page() && document().page()->hardwareKeyboardAttached())
+
+    RefPtr page = document().page();
+    if (optionsEnabled && page && page->chrome().client().isHardwareKeyboardAttached()) {
+        auto& client = document().page()->chrome().client();
+
+        // Registers a callback to exit fullscreen mode
+        client.startHardwareKeyboardAttachmentNotifications([weakThis = WeakPtr { *this }](bool attached) {
+            auto protectedThis = weakThis.get();
+            if (!protectedThis)
+                return;
+
+            // Exit the fullscreen API when the hardware keyboard is detached.
+            if (!attached) {
+                protectedThis->protectedDocument()->postTask(
+                    WTF::Function<void()>([weakThis]() {
+                        RefPtr protectedThis = weakThis.get();
+                        if (!protectedThis)
+                            return;
+
+                        RefPtr page = protectedThis->document().page();
+                        if (!page)
+                            return;
+
+                        auto& chromeClient = page->chrome().client();
+                        chromeClient.stopHardwareKeyboardAttachmentNotifications();
+
+                        Ref<Document> doc = protectedThis->protectedDocument().get();
+                        if (doc->fullscreen().fullscreenElement())
+                            doc->fullscreen().fullyExitFullscreen();
+                    })
+                );
+            }
```

Comment:
nit: I would do an early return
```suggestion
            if (attached)
                return;

            protectedThis->protectedDocument()->postTask(
                WTF::Function<void()>([weakThis]() {
                    RefPtr protectedThis = weakThis.get();
                    if (!protectedThis)
                        return;

                    RefPtr page = protectedThis->document().page();
                    if (!page)
                        return;

                    auto& chromeClient = page->chrome().client();
                    chromeClient.stopHardwareKeyboardAttachmentNotifications();

                    Ref<Document> doc = protectedThis->protectedDocument().get();
                    if (doc->fullscreen().fullscreenElement())
                        doc->fullscreen().fullyExitFullscreen();
                })
            );
```